pyocd/coresight/cortex_m.py

```diff
@@ -1304,3 +1328,272 @@ def exception_number_to_name(self, exc_num, name_thread=False):
     def in_thread_mode_on_main_stack(self):
         return (self._target_context.read_core_register('ipsr') == 0 and
                 (self._target_context.read_core_register('control') & CortexM.CONTROL_SPSEL) == 0)
+
+    def get_vector_catch_originating_stack_pointer_id(self):
+        # Reset vector catch has to be treated specially - LR is 0xFFFFFFFF in
+        # some circumstances, but not always, so can't rely on it. Instead check
+        # ipsr - if it says we're in Thread mode, then deduce it was a Reset,
+        # in which case we want to show MSP.
+        if self._target_context.read_core_register('ipsr') == 0:
+            return CortexM.MSP
+
+        if self._target_context.read_core_register('lr') & CortexM.EXC_RETURN_SPSEL:
+            return CortexM.PSP
+        else:
+            return CortexM.MSP
+
+#
+# Thread support
+#
+
+## @brief Class representing the handler mode.
+class MainStackThread(TargetThread):
```

Comment:
Could you move everything below here to a new module, maybe under `pyocd.debug`?